### PR TITLE
Immovable types prototype where the Move trait is builtin

### PR DESCRIPTION
**Pull request by Zoxc**

### CODE REVIEW

**Comment by aidanhs** [comment]
Crater run requested by zoxc/eddyb and now started.

File: src/libcore/cell.rs
Comment (plietar) [inline src/libcore/cell.rs]: These shouldn't be derived as the implementations would take pointers to the inner value.

**Comment by Zoxc** [comment]
This currently contains a hack where it catches all panics when building move data, a proper solution would be to ignore the errors and let borrowck handle them, see https://github.com/rust-lang/rust/issues/44830.

This also causes `FnOnce::Output` and `Deref::Target` to no longer implement `Move`, which is a breaking change.

**Comment by aidanhs** [comment]
At long last! Sorry for the delay, internal s3 errors delayed things a fair amount.

Cargobomb results: http://cargobomb-reports.s3.amazonaws.com/pr-44917/index.html. 'Blacklisted' crates (spurious failures etc) can be found
[here](https://github.com/rust-lang-nursery/cargobomb/blob/master/blacklist.md).
If you see any spurious failures not on the list, please make a PR against that file.

**Comment by Zoxc** [comment]
I looked over the results.

Errors due to duplicate bounds https://github.com/rust-lang/rust/issues/21974#issuecomment-331886186:
- nalgebra 0.10.1
- number_traits 0.1.2
- rowcol 0.3.3
- svgdom 0.7.0
- svgdom 0.6.0
- svgdom 0.3.1
- vek 0.2.2

The above errors is also expected to appear due to [the addition of the `DynSized` trait](https://github.com/rust-lang/rust/pull/44469).

Errors due to FnOnce::Output not implementing Move:
- parsell 0.6.5

Erros due to Deref::Target not implementing Move:
- shawshank 0.2.3
- thunk 0.3.0

Unknown errors:
- rust-install 0.0.4
- idcontain 0.7.2

**Comment by Zoxc** [comment]
I fixed the MoveData hack now that https://github.com/rust-lang/rust/pull/45016 finally landed.

**Comment by pnkfelix** [comment]
@Zoxc are you looking for review comments at this point?

**Comment by pnkfelix** [comment]
(@Zoxc to be clear, I am aware you've been working on this a while and discussing in the chat rooms. But I'm not sure if you expect the PR to actually be reviewed; as far as I know there isn't an RFC for this change, am I wrong?)

**Comment by pnkfelix** [comment]
@Zoxc having said that, I would like to at least coordinate the MIR-dataflow/MoveData changes made here so that they align with changes that I'm planning for the MoveData in MIR-borrowck.

**Comment by Zoxc** [comment]
@pnkfelix This is ready to be reviewed.

[There is an RFC](https://github.com/rust-lang/rfcs/pull/1858), although it was not been merged. This feature is required for experiments with immovable generators though (which does have a eRFC). The compiler team was also in favor of merging the `DynTrait` PR, which also lacks a RFC and has similar breakage.

File: src/libcore/cell.rs
Comment (arielb1) [inline src/libcore/cell.rs]: Can't a normal `Cell` take `T: ?Move`?

File: src/test/ui/on-unimplemented/multiple-impls.stderr
Comment (arielb1) [inline src/test/ui/on-unimplemented/multiple-impls.stderr]: Why are we losing errors here?

File: src/librustc/session/mod.rs
Comment (arielb1) [inline src/librustc/session/mod.rs]: is this still needed with `move_constraint`?

File: src/librustc/ty/mod.rs
Comment (arielb1) [inline src/librustc/ty/mod.rs]: I feel we should merge this function with `sized_constraint` and soon `dyn_sized_constraint`.

File: src/test/ui/on-unimplemented/multiple-impls.stderr
Comment (Zoxc) [reply src/test/ui/on-unimplemented/multiple-impls.stderr]: Good question. I also do not understand why there's duplicate errors here in the first place.

File: src/librustc/session/mod.rs
Comment (Zoxc) [reply src/librustc/session/mod.rs]: I don't think so. I reverted the changes to the recursion limit.

File: src/liballoc/boxed.rs
Comment (arielb1) [inline src/liballoc/boxed.rs]: Why do we have this requirement? Isn't this required by `FnOnce` for back-compat reasons?

File: src/librustc/lib.rs
Comment (arielb1) [inline src/librustc/lib.rs]: Is this needed with the new shortcutting?

File: src/librustc/traits/error_reporting.rs
Comment (arielb1) [inline src/librustc/traits/error_reporting.rs]: nit: preexisting, but I think
```Rust
if self.tcx.lang_items().move_trait() == Some(trait_def_id) ||
    self.tcx.lang_items().sized_trait() == Some(trait_def_id)
```

is the preferred style.

File: src/librustc/ty/mod.rs
Comment (arielb1) [reply src/librustc/ty/mod.rs]: See comment about merging this.

File: src/librustc/util/ppaux.rs
Comment (arielb1) [inline src/librustc/util/ppaux.rs]: nit: way too many newlines

File: src/librustc_mir/dataflow/move_paths/mod.rs
Comment (arielb1) [inline src/librustc_mir/dataflow/move_paths/mod.rs]: Of course it does, you're calling `all_paths_from::<&mut F>`, which further calls `all_paths_from::<&mut &mut F>` and so on. Please remove the commented code.

Comment (arielb1) [reply src/librustc_mir/dataflow/move_paths/mod.rs]: This is also a version of `on_all_children_bits` that does not ignore types within destructors. I really need to straighten the story there.

File: src/librustc_typeck/collect.rs
Comment (arielb1) [inline src/librustc_typeck/collect.rs]: nit: move the bounds into a struct rather than passing a `(bool, bool)`?

File: src/test/compile-fail/no-move-immovable-box-content.rs
Comment (arielb1) [inline src/test/compile-fail/no-move-immovable-box-content.rs]: Could you change this not to have an "use of possibly uninitialized variable" error, to prevent someone from refactoring this test away?

e.g., have a proper demonstration of unsoundness:
```Rust
fn move_from_box(b: Box<Immovable>) {
    *a; //~ ERROR cannot move immovable value out of a Box type
}

fn main() {
    let a = box Immovable;
    &*a; // ok
    move_from_box(a); // ok
}
```

File: src/test/ui/on-unimplemented/multiple-impls.stderr
Comment (arielb1) [reply src/test/ui/on-unimplemented/multiple-impls.stderr]: This needs to be investigated.

**Review by arielb1 — CHANGES_REQUESTED** [review]
I need to talk with @nikomatsakis to see what can be done about #21974.

Also, there are a few nits and disappearing errors are never a good thing, but otherwise r=me.

File: src/librustc/lib.rs
Comment (Zoxc) [reply src/librustc/lib.rs]: This is due to the increased amount of queries.

Comment (arielb1) [reply src/librustc/lib.rs]: Shouldn't the shortcutting avoid these?

File: src/liballoc/boxed.rs
Comment (Zoxc) [reply src/liballoc/boxed.rs]: It was just a bound to make things compile. I'll change FnBox instead.

File: src/test/compile-fail/no-move-immovable-box-content.rs
Comment (Zoxc) [reply src/test/compile-fail/no-move-immovable-box-content.rs]: You can't actually create boxes with immovable types yet. I do not want to litter the code with `#[cfg(stage0)]`.

File: src/librustc/lib.rs
Comment (Zoxc) [reply src/librustc/lib.rs]: I mean the macros which defines the queries causes the overflow. Not related to immovable types.

File: src/librustc/ty/mod.rs
Comment (Zoxc) [reply src/librustc/ty/mod.rs]: I'd like to delay that change until both `Move` and `DynTrait` PRs are merged.

File: src/librustc/traits/error_reporting.rs
Comment (Zoxc) [reply src/librustc/traits/error_reporting.rs]: Fixed.

File: src/librustc/util/ppaux.rs
Comment (Zoxc) [reply src/librustc/util/ppaux.rs]: Fixed.

File: src/librustc_mir/dataflow/move_paths/mod.rs
Comment (Zoxc) [reply src/librustc_mir/dataflow/move_paths/mod.rs]: I removed this.

File: src/librustc_typeck/collect.rs
Comment (Zoxc) [reply src/librustc_typeck/collect.rs]: I created a struct for this and also made `Bounds` and `compute_bounds` use it.

File: src/test/compile-fail/no-move-immovable-box-content.rs
Comment (arielb1) [reply src/test/compile-fail/no-move-immovable-box-content.rs]: Could you at least add a FIXME on the test?

Comment (Zoxc) [reply src/test/compile-fail/no-move-immovable-box-content.rs]: I changed the test to use your code.

File: src/libcore/cell.rs
Comment (bjorn3) [inline src/libcore/cell.rs]: missing newline

**Review by bjorn3 — CHANGES_REQUESTED** [review]

**Comment by ishitatsuyuki** [comment]
I think this hasn't undergo a RFC, so please allow me to ask what's an actual use case of this?

Move-safe is a great characteristic of Rust and I'm personally defensive to this.

**Comment by aidanhs** [comment]
@ishitatsuyuki the (unmerged) RFC is at https://github.com/rust-lang/rfcs/pull/1858, as mentioned above. The motivation section there is a good representation of why I personally want this.

**Comment by carols10cents** [comment]
Triage ping for you @Zoxc! How's it going?

**Comment by Zoxc** [comment]
I'm just waiting for https://github.com/rust-lang/rust/pull/45247 to land before rebasing this. Otherwise this PR is blocked on the breaking changes it causes.

**Comment by carols10cents** [comment]
> I'm just waiting for #45247 to land before rebasing this. Otherwise this PR is blocked on the breaking changes it causes.

Triage ping for you @Zoxc since #45247 has been merged!

**Comment by kennytm** [comment]
@arielb1 @pnkfelix in your court again

**Comment by carols10cents** [comment]
Friendly ping to keep this on your radar @pnkfelix !

File: src/libcore/marker.rs
Comment (pnkfelix) [inline src/libcore/marker.rs]: weird: do you have to have `?Move` constraint on this trait?

Just curious, since I noticed that `trait Sized` does not have an analogous `?Sized` bound...

File: src/librustc_passes/ast_validation.rs
Comment (pnkfelix) [inline src/librustc_passes/ast_validation.rs]: You've removed all the calls to this method. I infer that it because you want to assume that traits automatically start off as implementing `Move`, and thus you need to be able to put `: ?Move` as bound on trait definitions in order to override that initial default.

But shouldn't we replace the old function with one that is specialized to `?Sized`? Or is there some reason why we want to stop erroring in response to something like `trait X : ?Sized { fn x(&self); }` ?

Comment (pnkfelix) [reply src/librustc_passes/ast_validation.rs]: oh never mind, I see that you did add such a check in `fn conv_object_ty_poly_trait_ref`

File: src/test/compile-fail/maybe-bounds.rs
Comment (pnkfelix) [inline src/test/compile-fail/maybe-bounds.rs]: hmm... I'm a little worried that this error has gone away. (I'd need to review the details about why its an error in the first place, but assuming that there was a good reason for it being an error, I assume that those same reason still hold here? Or was this just a case of "trait objects are implicitly unsized and thus it makes no sense to put `?Sized` in them" ?

File: src/test/compile-fail/move-trait-leaks-in-impl-trait.rs
Comment (pnkfelix) [inline src/test/compile-fail/move-trait-leaks-in-impl-trait.rs]: hmm. This is certainly fine as a conservative choice for now (I do see that the RFC says "Existential types (impl Trait) are Move if their underlying type are Move" which I take to be an "if *and only if*").

But I find it interesting because I would have guessed that since no borrows are taken of the value returned from `immovable()`, nor do we call any methods on it that take `&self`/`&mut self`, that it would be safe to subsequently move it in the call to `test(..)`

File: src/test/compile-fail/phantom-oibit.rs
Comment (pnkfelix) [inline src/test/compile-fail/phantom-oibit.rs]: Did ... you need to add this? In other words, was it injecting new compiler errors otherwise? (If so, what were they?)

If not, then I assume you just added it because its the right way to preserve the previous semantics of the test here?

File: src/test/compile-fail/privacy-sanity.rs
Comment (pnkfelix) [inline src/test/compile-fail/privacy-sanity.rs]: (likewise here, I am again curious as to whether this was a necessary change, or just an artifact of your methodology for updating the tests...)

Comment (pnkfelix) [reply src/test/compile-fail/privacy-sanity.rs]: (I'll stop asking about the future instances of this pattern now, though.)

File: src/test/ui/impl-trait/equality-recursive.rs
Comment (pnkfelix) [inline src/test/ui/impl-trait/equality-recursive.rs]: I see you factored this test out from a separate piece of code that started failing (in a new way) in `src/test/ui/impl-trait/equality.rs` ... do we need to be concerend about this change in behavior? I.e. Does it need a separate issue filed for it, once this PR lands?

**Comment by pnkfelix** [comment]
Looks fine to me.

I had a bunch of questions that I posted as comments on the diff. But I don't think any of them actually warrant holding up trying to land this (assuming that we're collectively fine with landing this as an experimental extension for non-finalized RFC).

So, r=me after a rebase (but I'd love to see those questions answered!)

File: src/test/compile-fail/no-move-immovable-box-content.rs
Comment (bjorn3) [inline src/test/compile-fail/no-move-immovable-box-content.rs]: Missing newline

File: src/libstd/panic.rs
Comment (bjorn3) [inline src/libstd/panic.rs]: Missing spaces around the `+`

File: src/libstd/panic.rs
Comment (bjorn3) [inline src/libstd/panic.rs]: Missing spaces around the `+`

File: src/test/ui/impl-trait/equality-recursive.rs
Comment (bjorn3) [inline src/test/ui/impl-trait/equality-recursive.rs]: Missing newline

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/test/compile-fail/maybe-bounds.rs
Comment (Zoxc) [reply src/test/compile-fail/maybe-bounds.rs]: Trait objects previously had no default bounds, so the `dyn TraitA + ?TraitB` syntax wasn't useful. However now you can use it with `Move`, so `dyn Trait` implements `Move`, `dyn Trait + ?Move` does not.

File: src/test/compile-fail/move-trait-leaks-in-impl-trait.rs
Comment (Zoxc) [reply src/test/compile-fail/move-trait-leaks-in-impl-trait.rs]: It is safe to move it into `test`. The problem is that the `test` function has a `Move` bound, so it may rely on the type being movable after taking the address of it.

File: src/libcore/marker.rs
Comment (Zoxc) [reply src/libcore/marker.rs]: Auto traits cannot have supertraits because that is unsound. Also we do not want `Move` to require `Move`. That would probably result in interesting things happening.

File: src/test/compile-fail/phantom-oibit.rs
Comment (Zoxc) [reply src/test/compile-fail/phantom-oibit.rs]: Auto traits cannot have supertraits because that is unsound, so we need to opt-out to `Move`, which is a default bound on `Self` for all traits.

File: src/test/ui/impl-trait/equality-recursive.rs
Comment (Zoxc) [reply src/test/ui/impl-trait/equality-recursive.rs]: I split this out since this error happens in another compiler phase compared to the rest. I don't remember the details, but I probably did something which caused the compiler to exit earlier due to errors.

**Comment by Zoxc** [comment]
I'm not sure why @pnkfelix is still assigned here. I saw @arielb1 unassign him in the log. Anyway this is still blocked on @arielb1 and @nikomatsakis looking into fixing the breaking changes and looking into the changes on the `src/test/ui/on-unimplemented/multiple-impls.stderr` test.

**Comment by kennytm** [comment]
Hi @Zoxc, is this still blocked by the `src/test/ui/on-unimplemented/multiple-impls.stderr` test and the breaking changes?

**Comment by Zoxc** [comment]
Yes.

**Comment by carols10cents** [comment]
Friendly ping @arielb1 and @nikomatsakis, this comment seems to say this is waiting on you!

> Anyway this is still blocked on @arielb1 and @nikomatsakis looking into fixing the breaking changes and looking into the changes on the src/test/ui/on-unimplemented/multiple-impls.stderr test.

**Comment by kennytm** [comment]
Triage — this PR appears to be still blocked by 

> Anyway this is still blocked on @arielb1 and @nikomatsakis looking into fixing the breaking changes and looking into the changes on the `src/test/ui/on-unimplemented/multiple-impls.stderr` test.

BTW @Zoxc There is merge conflict.

**Comment by arielb1** [comment]
## Status of the Move/DynSized PRs

I apologize that this was not communicated clearly enough, but there are breakage and language-design issues blocking these PRs:

1. First, there's a trait-system inference issue that means that without NLL and another fix, this PR "semi-randomly" breaks user code, which means that, as both NLL and not causing unnecessary breakage are fairly high priority, and these PRs are less so, we should probably delay this PR until NLL stabilizes.
     We only discovered this when we have done the crater run on the `Move` PR. Thanks @Zoxc for getting to this state where this is our worst technical issue.
2. Second, as you have seen, adding more "Sized-style" traits has an ergonomics disadvantage which we didn't appreciate enough when we have first seen this PR.
     We (or rather the lang team) need to find the best compromise in this area, which I'm not that sure about. Feel free to join them on this issue - they should have more time after the holidays, now that the impl period is over.

**Comment by pnkfelix** [comment]
https://github.com/rust-lang/rfcs/issues/2255 seems relevant to @arielb1's second point (quoted here)

> Second, as you have seen, adding more "Sized-style" traits has an ergonomics disadvantage which we didn't appreciate enough when we have first seen this PR.
We (or rather the lang team) need to find the best compromise in this area, which I'm not that sure about. Feel free to join them on this issue - they should have more time after the holidays, now that the impl period is over.

**Pull request closed without merging**
